File: solutions/csharp/matching-brackets/MatchingBrackets.cs
public static class MatchingBrackets
{
    public static bool IsPaired(string s)
    {
        Stack<char> stack = new();

        foreach (char c in s)
        {
            if (c == '(' || c == '[' || c == '{')
            {
                stack.Push(c);
            }
            else if (c == ')')
            {
                if (stack.Count == 0) return false;
                if (stack.Peek() != '(')
                    return false;
                else
                    stack.Pop();
            }
            else if (c == ']')
            {
                if (stack.Count == 0) return false;
                if (stack.Peek() != '[')
                    return false;
                else
                    stack.Pop();
            }
            else if (c == '}')
            {
                if (stack.Count == 0) return false;
                if (stack.Peek() != '{')
                    return false;
                else
                    stack.Pop();
            }
        }

        return stack.Count == 0;
    }
}

File: solutions/python/matching-brackets/matching_brackets.py
def is_paired(input_string):

    stack = []

    for c in input_string:
        if c == "(" or c == "[" or c == "{":
            stack.append(c)
        elif c == ")":
            if stack == []:
                return False
            if stack[-1] != "(":
                return False
            else:
                stack.pop()
        elif c == "]":
            if stack == []:
                return False
            if stack[-1] != "[":
                return False
            else:
                stack.pop()
        elif c == "}":
            if stack == []:
                return False
            if stack[-1] != "{":
                return False
            else:
                stack.pop()

    return stack == []
